File: 2025-2/sesion-de-clase-10/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/inf30/softprog/daoimpl/ventas/OrdenVentaDAOImpl.java
package pe.edu.pucp.inf30.softprog.daoimpl.ventas;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.dao.ventas.OrdenVentaDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.TransaccionalBaseDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.clientes.ClienteDAOImpl;
import pe.edu.pucp.inf30.softprog.daoimpl.rrhh.EmpleadoDAOImpl;
import pe.edu.pucp.inf30.softprog.modelo.ventas.OrdenVenta;

/**
 *
 * @author eric
 */
public class OrdenVentaDAOImpl extends TransaccionalBaseDAO<OrdenVenta> 
        implements OrdenVentaDAO {

    @Override
    protected PreparedStatement comandoCrear(Connection conn, OrdenVenta modelo) 
            throws SQLException {
        
        String sql = "{call insertarOrdenVenta(?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        
        if (modelo.getCliente() != null) {
            cmd.setInt("p_idCliente", modelo.getCliente().getId());
        }
        else {
            cmd.setNull("p_idCliente", Types.INTEGER);
        }
        
        
        if (modelo.getEmpleado()!= null) {
            cmd.setInt("p_idEmpleado", modelo.getEmpleado().getId());
        }
        else {
            cmd.setNull("p_idEmpleado", Types.INTEGER);
        }
        
        cmd.setDouble("p_total", modelo.getTotal());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, 
            OrdenVenta modelo) throws SQLException {
        
        String sql = "{call modificarOrdenVenta(?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        
        if (modelo.getCliente() != null) {
            cmd.setInt("p_idCliente", modelo.getCliente().getId());
        }
        else {
            cmd.setNull("p_idCliente", Types.INTEGER);
        }
        
        
        if (modelo.getEmpleado()!= null) {
            cmd.setInt("p_idEmpleado", modelo.getEmpleado().getId());
        }
        else {
            cmd.setNull("p_idEmpleado", Types.INTEGER);
        }
        
        cmd.setDouble("p_total", modelo.getTotal());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.setInt("p_id", modelo.getId());
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, 
            Integer id) throws SQLException {
        
        String sql = "{call eliminarOrdenVenta(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, 
            Integer id) throws SQLException {
        
        String sql = "{call buscarOrdenVentaPorId(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) 
            throws SQLException {
        
        String sql = "{call listarOrdenesVenta()}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        return cmd;
    }

    @Override
    protected OrdenVenta mapearModelo(ResultSet rs) throws SQLException {
        OrdenVenta ordenVenta = new OrdenVenta();
        ordenVenta.setId(rs.getInt("id"));
        
        int idEmpleado = rs.getInt("idEmpleado");
        if (!rs.wasNull()) {
            ordenVenta.setEmpleado(
                new EmpleadoDAOImpl().leer(idEmpleado));
        }
        
        int idCliente = rs.getInt("idCliente");
        if (!rs.wasNull()) {
            ordenVenta.setCliente(
                new ClienteDAOImpl().leer(idCliente));
        }
        
        ordenVenta.setTotal(rs.getDouble("total"));
        ordenVenta.setActivo(rs.getBoolean("activo"));
        
        return ordenVenta;
    }
}

File: 2025-2/sesion-de-clase-15/SoftProgPersistencia/DAOImpl/Ventas/OrdenVentaDAOImpl.cs
using System;
using System.Data;
using System.Data.Common;
using PUCP.SoftProg.Modelo.Ventas;
using PUCP.SoftProg.Persistencia.DAO.Ventas;
using PUCP.SoftProg.Persistencia.DAOImpl.Clientes;
using PUCP.SoftProg.Persistencia.DAOImpl.RRHH;

namespace PUCP.SoftProg.Persistencia.DAOImpl.Ventas {
    public class OrdenVentaDAOImpl : TransactionalBaseDAOImpl<OrdenVenta>, IOrdenVentaDAO {
        protected override DbCommand ComandoCrear(DbConnection conn, OrdenVenta ordenVenta) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarOrdenVenta";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCliente", DbType.Int32, ordenVenta.Cliente.Id);
            this.AgregarParametroEntrada(cmd, "@p_idEmpleado", DbType.Int32, DBNull.Value);
            this.AgregarParametroEntrada(cmd, "@p_total", DbType.Double, ordenVenta.Total);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, ordenVenta.IsActive);
            this.AgregarParametroSalida(cmd, "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand ComandoActualizar(DbConnection conn, OrdenVenta ordenVenta) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarOrdenVenta";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCliente", DbType.Int32, ordenVenta.Cliente.Id);
            this.AgregarParametroEntrada(cmd, "@p_idEmpleado", DbType.String, ordenVenta.Empleado.Id);
            this.AgregarParametroEntrada(cmd, "@p_total", DbType.String, ordenVenta.Total);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, ordenVenta.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, ordenVenta.Id);

            return cmd;
        }

        protected override DbCommand ComandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarOrdenVenta";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeer(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarOrdenVentaPorId";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeerTodos(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarOrdenesVenta";
            cmd.CommandType = CommandType.StoredProcedure;
            return cmd;
        }

        protected override OrdenVenta MapearModelo(DbDataReader reader) {
            return new OrdenVenta {
                Id = Convert.ToInt32(reader["id"]),
                Cliente = new ClienteDAOImpl().Leer(Convert.ToInt32(reader["idCliente"])), 
                Empleado = new EmpleadoDAOImpl().Leer(Convert.ToInt32(reader["idEmpleado"])), 
                Total = Convert.ToInt32(reader["total"]),
                IsActive = Convert.ToBoolean(reader["activo"])
            };
        }
    }
}
